<!DOCTYPE HTML>
<html>

<head>
  <title>Publication | Ho Lab </title>
  <meta name="description" content="This laboratory, led by Dr. Joshua Ho, focuses on the use of bioinformatics and systems biology approaches to tackle longstanding problems in basic biology and translational medicine." />
  <meta name="keywords" content="Bioinformatics, Systems Medicine, Systems Biology, Joshua Ho, University of Hong Kong, HKU, School of Biomedical Sciences, single cell, single-cell RNA-seq, developmental biology, artificial intelligence, big data, epigenomics, whole genome sequencing, next-generation sequencing, Gene regulatory network" />
  <meta http-equiv="content-type" content="text/html; charset=windows-1252" />
  <link rel="stylesheet" type="text/css" href="style/style.css" />
</head>


<body>
  <div id="main">
    <div id="header">
      <div id="logo">
        <div id="logo_text">
          <!-- class="logo_colour", allows you to change the colour of the text -->
          <h1><a href="index.html"><span class="logo_colour">Bioinformatics and Systems Medicine</span></a></h1>
          <h2>School of Biomedical Sciences, Li Ka Shing Faculty of Medicine, The University of Hong Kong</h2>
        </div>
      </div>

      <div id="menubar">
      <ul id="menu">
          <!-- put class="selected" in the li tag for the selected page - to highlight which page you're on -->
          <li><a href="index.html">Home</a></li>
          <li><a href="people.html">People</a></li>
          <li><a href="research.html">Research</a></li>
          <li><a href="publication.html">Publication</a></li>
          <li class="selected"><a href="software.html">Software</a></li>
          <li><a href="contact.html">Contact</a></li>
        </ul>
      </div>
    </div>
    <div id="content_header"></div>
    <div id="site_content">
      <div id="content">
        <!-- insert the page content here -->
      <h1 class="page-header">Software</h1>
            <ul>
	      <li><a href="https://github.com/VCCRI/CIDR">CIDR</a> is an R package thay implements an ultrafast and accurate dimensionality reduction and clustering algorithm for single-cell RNA-seq data. The method is described in this <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-017-1188-0"><i>Genome Biology</i></a> paper.</li>
	      <li><a href="https://github.com/VCCRI/Falco">Falco</a> is flexible big data framework for fast processing of single-cell RNA-seq data on the cloud. The method is described in this <a href="http://bioinformatics.oxfordjournals.org/content/early/2016/12/24/bioinformatics.btw732.abstract"><i>Bioinformatics</i></a> paper.</li>
	      <li><a href="https://github.com/VCCRI/PBrowse">PBrowse</a> is web-based platform for real-time collaborative exploration and sharing of genomic data. A test server can be found <a href="http://pbrowse.victorchang.edu.au/">here</a>. The method is described in this <a href="http://dx.doi.org/10.1093/nar/gkw1358"><i>Nucleic Acids Research</i></a> paper.</li>	
	      <li><a href="https://github.com/VCCRI/PAD">PAD</a> is a web-based bioinformatics tool for analysing transcription factor (TF) co-binding at gene-proximal or distal regulatory elements in mouse embryonic stem cells. A test server can be found <a href="http://pad.victorchang.edu.au/">here</a>. The method is described in this <a href="https://academic.oup.com/bioinformatics/article-lookup/doi/10.1093/bioinformatics/btx092"><i>Bioinformatics</i></a> paper. </li>
 	      <li><a href="https://github.com/VCCRI/XGSA">XGSA</a> is an R package that implements our statistical method for cross-species gene set analysis. A full description of the method can be found in this <a href="http://bioinformatics.oxfordjournals.org/content/32/17/i620.full"><i>Bioinformatics</i> paper</a>.  	
              <li><a href="http://modencode.med.harvard.edu/data_sets/chromatin/">ENCODE-X Cross-species chromatin data browser</a> is an online portal for ENCODE and modENCODE chromatin datasets. A full description of the data can be found in our <a href="http://www.nature.com/nature/journal/v512/n7515/full/nature13415.html"=><i>Nature</i> paper</a>.</li>
              <li><a href="https://github.com/kasohn/hiHMM">hiHMM</a> is a new Bayesian non-parametric method to jointly infer chromatin state maps in multiple genomes (different cell types, developmental stages, even multiple species) using genome-wide histone modification data.</li>
              <li><a href="http://compbio.med.harvard.edu/ToothCODE/">ToothCODE</a> is an online portal of embryonic mouse tooth development data generated by the tooth regeneration team of SysCODE.</li>
              <li><a href="http://bioinformatics.udel.edu/Research/iSyTE/">iSyTE</a> is an integrated systems tool for eye gene discovery.</li>
              <li><a href="https://github.com/hojwk/SeqVis">SeqVis</a> is a stand-alone, platform-independent Java application developed with the aim to facilitate analysis and 3D visualization of compositional heterogeneity in species-rich alignments of nucleotide sequences. A software website can be found <a href="http://sydney.edu.au/science/biology/about_us/honorary_staff/jermiin_lars/seqvis/index.htm">here</a> and <a href="http://bioinformatics.victorchang.edu.au/seqvis/">here</a></li>
              <li><a href="tools/DVA_2.0.zip">DVA</a> is a R script that performs differential variability analysis. A full description of this method can be found in this <a href="http://bioinformatics.oxfordjournals.org/content/24/13/i390.full"><i>Bioinformatics</i> paper</a>.
              <li><a href="https://github.com/VCCRI/C3/">C3</a> is an R package for cross-species compendium-based cell-type identification.</li>
              <li><a href="https://github.com/VCCRI/GEOracle">GEOracle</a> is an R/shiny package for identifying and analysing perturbation experiments in NCBI GEO. Source code can be found in <a href="https://github.com/VCCRI/GEOracle">GitHub</a>.</li>
              <li><a href="https://github.com/VCCRI/Scavenger/">Scavenger</a> a phyton tool to recover false negative unmapped reads.</li>
              <li><a href="ularcirc.html">Ularcirc</a> is an R package for visualisation and analysis of circular RNA.</li>
              <li><a href="https://github.com/VCCRI/CardiacProfileR">CardiacProfileR</a> is an R package for extraction and visualisation of heart rate profiles from wearable fitness trackers.</li>
              <li><a href="https://github.com/VCCRI/starmap">starmap</a> is a web-based VR-enabled tool which combines a 3D scatter plot with star plots to visualise hundreds of thousands of multivariate data points, such as single-cell expression data. starmap can be accessed from a desktop, laptop or a mobile device from the following link: <a href="https://vccri.github.io/starmap/">https://vccri.github.io/starmap/</a>.</li>

              <li><a href="https://github.com/VCCRI/FlowGrid">FlowGrid</a> is an ultrafast clustering algorithm for single-cell flow cytometry data. It implements a novel grid-based DBSCAN clustering algorithm.</li>

 </ul>
            <h1 class="page-header">Data sets</h1>
            <ul>
              <li>Antibody microarray of patients with Systemic Lupus Erythematosus (<a href="http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?&acc=GSE27293">GEO</a>)</li>
              <li>Histone modification and gene expression in dosage compensation of <i>Drosophila melanogaster</i> (<a href="http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE34859">GEO</a>)</li>
              <li>Developmental timecourse of mouse ocular lens and whole embryo control (<a href="http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE32334">GEO</a>)</li>
              <li>Gene expression during embryonic mouse molar tooth development (<a href="http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE32321">GEO</a>)</li>
            </ul>


      </div>
    </div>
   <div id="content_footer"></div>
     <div id="content_footer"></div>
     <div id="footer">
       Copyright &copy; <a href="http://holab-hku.github.io/" target="_blank">Joshua Ho</a>
    </div>
    <p>&nbsp;</p>
  </div>
</body>
</html>
