(this["webpackJsonpholab-website"]=this["webpackJsonpholab-website"]||[]).push([[0],{29:function(e,a,t){e.exports=t(58)},34:function(e,a,t){},35:function(e,a,t){},58:function(e,a,t){"use strict";t.r(a);var n=t(0),l=t.n(n),r=t(26),i=t.n(r),c=(t(34),t(35),t(1)),o=t(7);var s=function(){return l.a.createElement("div",{id:"site_content"},l.a.createElement("div",{id:"sidebar_container"},l.a.createElement("div",{className:"sidebar"},l.a.createElement("div",{className:"sidebar_top"}),l.a.createElement("div",{className:"sidebar_item"},l.a.createElement("h4",null,"Big Data special issue published!"),l.a.createElement("h5",null,"25 February 2019"),l.a.createElement("p",null,l.a.createElement("a",{href:"https://link.springer.com/journal/12551/11/1"},l.a.createElement("img",{src:"img/Cover.jpg",alt:"",width:"110"})))),l.a.createElement("div",{className:"sidebar_base"}))),l.a.createElement("div",{id:"content0"},l.a.createElement("h3",null,"Welcome to the laboratory of Dr. Joshua Ho"),l.a.createElement(c.b,{to:"/research"}," ",l.a.createElement("img",{src:"img/wordle.PNG",alt:"",width:"600"})," ")))};var m=function(){return l.a.createElement("div",{id:"site_content"},l.a.createElement("div",{id:"content"},l.a.createElement("div",null,l.a.createElement("h3",null,l.a.createElement("b",null,"Dr Joshua W. K. Ho")," - Laboratory Head"),l.a.createElement(c.b,{to:"/people/JoshuaHo"},l.a.createElement("img",{src:"img/JoshuaHo.jpg",width:"100",align:"left",alt:"",style:{margin:"0px 15px 0px 0px"}})," "),l.a.createElement("p",null,l.a.createElement("b",null,"BSc (Hon 1, University Medal)"),", Biochemistry and Computer Science, The University of Sydney, Australia",l.a.createElement("br",null),l.a.createElement("b",null,"PhD"),", Bioinformatics, The University of Sydney, Australia",l.a.createElement("br",null),l.a.createElement(c.b,{to:"/people/JoshuaHo"},"Find out more..."))),l.a.createElement("br",null),l.a.createElement("div",null,l.a.createElement("h3",null,"Current members"),l.a.createElement("b",null,"Dr Ken Yu")," (postdoctoral fellow)",l.a.createElement("br",null),l.a.createElement("b",null,"Dr Haobin Yao")," (postdoctoral fellow)",l.a.createElement("br",null),l.a.createElement("b",null,"Xiunan Fang")," (PhD student)",l.a.createElement("br",null),l.a.createElement("b",null,"Gordon Qian")," (PhD student)",l.a.createElement("br",null)),l.a.createElement("div",null,l.a.createElement("h3",null,"Alumni"),l.a.createElement("a",{href:"https://www.victorchang.edu.au/heart-research/computational-genomics"},l.a.createElement("b",null,"Dr Eleni Giannoulatou"))," (senior postdoc 2013-2015 at VCCRI; Now Division Head at VCCRI)",l.a.createElement("br",null),l.a.createElement("b",null,"Dr Paul Lin")," (postdoc 2016-2017 at VCCRI; Now technical manager at the Independent Hospital Pricing Authority, Australia)",l.a.createElement("br",null),l.a.createElement("b",null,"Dr Djordje Djordjevic")," (PhD student 2014-2017 at UNSW; Now Computational Biologist at Novo Nordisk, Denmark)",l.a.createElement("br",null),l.a.createElement("b",null,"Dr Xin Wang")," (PhD student 2014-2018 at UNSW; Now postdoc at Northwestern University, USA)",l.a.createElement("br",null),l.a.createElement("b",null,"Dr Tomasz Szczesnik")," (PhD student 2014-2018 at UNSW; Now postdoc at ETH Zurich, Switzerland)",l.a.createElement("br",null),l.a.createElement("b",null,"Dr Andrian Yang")," (PhD student 2015-2018 at UNSW; Now postdoc at EMBL-EBI, Cambridge, UK)",l.a.createElement("br",null))))};var u=function(){return l.a.createElement("div",{id:"site_content"},l.a.createElement("div",{id:"content"},l.a.createElement("h1",null,"Research projects"),l.a.createElement("p",null,"The Ho Laboratory focuses on the use of bioinformatics and systems biology approaches to tackle longstanding problems in basic and translational medicine. A range of specific PhD projects can be developed within the broad theme of ",l.a.createElement("b",null,"scalable big data analytics for healthcare translation"),". Here are some major research themes. Multiple projects are available under each theme.:"),l.a.createElement("ul",null,l.a.createElement("li",null,l.a.createElement("b",null,"Scalable single cell data analytics"),". Single-cell RNA sequencing (scRNA-Seq) enables researchers to study heterogeneity among tens of thousands of individual cells and define cell types from a transcriptomic perspective. However, fast and reliable analysis of these large and noisy data requires new statistical and computational considerations. In this project we will develop scalable bioinformatics methods to analyze a range of scRNA-seq data to answer important biological questions.",l.a.createElement("br",null),l.a.createElement("br",null)),l.a.createElement("li",null,l.a.createElement("b",null,"Microbiome functional systems biology through metagenomic and multi-omic data analysis"),". Our laboratory is developing computational and statistical tools that can efficient process large metagenomic data, and integrate them with other omics or deep phenotyping data. Our goal is to understand how the microbiome found in specific location of the body, e.g., the gut, can affect a person's health.",l.a.createElement("br",null),l.a.createElement("br",null)),l.a.createElement("li",null,l.a.createElement("b",null,"Medical artificial intelligence, mobile health and wearable devices"),". Being able to track the changes of a person's physiological parameters in real time is now increasingly feasible due to the wide availability of consumer-grade smartphones and wearable devices (e.g., fitbit, AppleWatch, etc). Our group is developing new big data machine-learning algorithms extract, de-noise, analyze and correlate physical activity data and heart rate dynamics. Our long-term goal is to establish new non-invasive screening tools to monitor a person's health status.",l.a.createElement("br",null),l.a.createElement("br",null))),l.a.createElement("p",null,l.a.createElement("b",null,"For postdoc/students/RA who wants to join this laboratory:")," All projects require proficiency in at least one programming/scripting language (R, Perl, Python, Java, C++, C) Familiarity with the Unix operating system is desirable but not required. Individual project can be tailored to fit each student's personal interest and skill set. Most projects involve close interactions with local and international collaborators. This is a highly interdisciplinary laboratory. We welcome perspective group members from diverse background, such as medicine, biology, physics, computer science, mathematics, statistics, and engineering. Expression of interest, along with your CV, can be sent to ",l.a.createElement(c.b,{to:"/contact"},"Dr. Ho"),".")))},h=t(10),d=t(9),E=t.n(d),p=t(13),b=t.n(p);var f=function(){var e=Object(n.useState)([]),a=Object(h.a)(e,2),t=a[0],r=a[1],i=Object(n.useState)([]),c=Object(h.a)(i,2),o=c[0],s=c[1],m=Object(n.useState)([]),u=Object(h.a)(m,2),d=u[0],p=u[1],b=Object(n.useState)([]),f=Object(h.a)(b,2),g=f[0],v=f[1];return Object(n.useEffect)((function(){E.a.get("/data/publication/ejsi.txt").then((function(e){var a=e.data;r(a.split("\n"))})),E.a.get("/data/publication/ppm.txt").then((function(e){var a=e.data;s(a.split("\n"))})),E.a.get("/data/publication/bc.txt").then((function(e){var a=e.data;p(a.split("\n"))})),E.a.get("/data/publication/orprp.txt").then((function(e){var a=e.data;v(a.split("\n"))}))}),[]),l.a.createElement("div",{id:"site_content"},l.a.createElement("div",{id:"content"},l.a.createElement("h1",null,"Publication"),l.a.createElement("p",null," ",l.a.createElement("a",{href:"http://scholar.google.com/citations?user=3kVPIhQAAAAJ&hl=en&oi=ao",target:"_blank",rel:"noopener noreferrer"},"Google Scholar")," \xa0\xa0\xa0 ",l.a.createElement("a",{href:"http://www.ncbi.nlm.nih.gov/pubmed/?term=Joshua+W+Ho%5BAuthor%5D",target:"_blank",rel:"noopener noreferrer"},"PubMed")," "),l.a.createElement("h2",null,"Edited journal special issue"),l.a.createElement("ol",{className:"semi-spaced-list"},l.a.createElement("a",{href:"https://link.springer.com/journal/12551/11/1"},l.a.createElement("img",{src:"img/Cover.jpg",width:"80",alt:""})),l.a.createElement("br",null),l.a.createElement("br",null),t.map((function(e,a){return l.a.createElement("li",{key:a,dangerouslySetInnerHTML:{__html:e}})}))),l.a.createElement("p",null,l.a.createElement("strong",null,"[* Co-first authors, ^ Co-corresponding authors]")),l.a.createElement("h2",null,"Pre-print manusripts"),l.a.createElement("ol",{className:"semi-spaced-list"},o.map((function(e,a){return l.a.createElement("li",{key:a,dangerouslySetInnerHTML:{__html:e}})}))),l.a.createElement("h2",null,"Book chapters"),l.a.createElement("ol",{className:"semi-spaced-list"},d.map((function(e,a){return l.a.createElement("li",{key:a,dangerouslySetInnerHTML:{__html:e}})}))),l.a.createElement("h2",null,"Original research papers / review papers"),l.a.createElement("ol",{className:"semi-spaced-list"},g.map((function(e,a){return l.a.createElement("li",{key:a,dangerouslySetInnerHTML:{__html:e}})})))))};var g=function(){return l.a.createElement("div",{id:"site_content"},l.a.createElement("div",{id:"content"},l.a.createElement("h1",{className:"page-header"},"Software"),l.a.createElement("ul",null,l.a.createElement("li",null,l.a.createElement("a",{href:"https://github.com/VCCRI/CIDR"},"CIDR")," is an R package thay implements an ultrafast and accurate dimensionality reduction and clustering algorithm for single-cell RNA-seq data. The method is described in this ",l.a.createElement("a",{href:"https://genomebiology.biomedcentral.com/articles/10.1186/s13059-017-1188-0"},l.a.createElement("i",null,"Genome Biology"))," paper."),l.a.createElement("li",null,l.a.createElement("a",{href:"https://github.com/VCCRI/Falco"},"Falco")," is flexible big data framework for fast processing of single-cell RNA-seq data on the cloud. The method is described in this ",l.a.createElement("a",{href:"http://bioinformatics.oxfordjournals.org/content/early/2016/12/24/bioinformatics.btw732.abstract"},l.a.createElement("i",null,"Bioinformatics"))," paper."),l.a.createElement("li",null,l.a.createElement("a",{href:"https://github.com/VCCRI/PBrowse"},"PBrowse")," is web-based platform for real-time collaborative exploration and sharing of genomic data. A test server can be found ",l.a.createElement("a",{href:"http://pbrowse.victorchang.edu.au/"},"here"),". The method is described in this ",l.a.createElement("a",{href:"http://dx.doi.org/10.1093/nar/gkw1358"},l.a.createElement("i",null,"Nucleic Acids Research"))," paper."),l.a.createElement("li",null,l.a.createElement("a",{href:"https://github.com/VCCRI/PAD"},"PAD")," is a web-based bioinformatics tool for analysing transcription factor (TF) co-binding at gene-proximal or distal regulatory elements in mouse embryonic stem cells. A test server can be found ",l.a.createElement("a",{href:"http://pad.victorchang.edu.au/"},"here"),". The method is described in this ",l.a.createElement("a",{href:"https://academic.oup.com/bioinformatics/article-lookup/doi/10.1093/bioinformatics/btx092"},l.a.createElement("i",null,"Bioinformatics"))," paper. "),l.a.createElement("li",null,l.a.createElement("a",{href:"https://github.com/VCCRI/XGSA"},"XGSA")," is an R package that implements our statistical method for cross-species gene set analysis. A full description of the method can be found in this ",l.a.createElement("a",{href:"http://bioinformatics.oxfordjournals.org/content/32/17/i620.full"},l.a.createElement("i",null,"Bioinformatics")," paper"),"."),l.a.createElement("li",null,l.a.createElement("a",{href:"http://modencode.med.harvard.edu/data_sets/chromatin/"},"ENCODE-X Cross-species chromatin data browser")," is an online portal for ENCODE and modENCODE chromatin datasets. A full description of the data can be found in our ",l.a.createElement("a",{href:"http://www.nature.com/nature/journal/v512/n7515/full/nature13415.html"},l.a.createElement("i",null,"Nature")," paper"),"."),l.a.createElement("li",null,l.a.createElement("a",{href:"https://github.com/kasohn/hiHMM"},"hiHMM")," is a new Bayesian non-parametric method to jointly infer chromatin state maps in multiple genomes (different cell types, developmental stages, even multiple species) using genome-wide histone modification data."),l.a.createElement("li",null,l.a.createElement("a",{href:"http://compbio.med.harvard.edu/ToothCODE/"},"ToothCODE")," is an online portal of embryonic mouse tooth development data generated by the tooth regeneration team of SysCODE."),l.a.createElement("li",null,l.a.createElement("a",{href:"http://bioinformatics.udel.edu/Research/iSyTE/"},"iSyTE")," is an integrated systems tool for eye gene discovery."),l.a.createElement("li",null,l.a.createElement("a",{href:"https://github.com/hojwk/SeqVis"},"SeqVis")," is a stand-alone, platform-independent Java application developed with the aim to facilitate analysis and 3D visualization of compositional heterogeneity in species-rich alignments of nucleotide sequences. A software website can be found ",l.a.createElement("a",{href:"http://sydney.edu.au/science/biology/about_us/honorary_staff/jermiin_lars/seqvis/index.htm"},"here")," and ",l.a.createElement("a",{href:"http://bioinformatics.victorchang.edu.au/seqvis/"},"here")),l.a.createElement("li",null,l.a.createElement("a",{href:"https://github.com/VCCRI/SeqVis"},"SeqVis 2.0")," is a web-based version of SeqVis. The demo can be found here: ",l.a.createElement("a",{href:"https://vccri.github.io/SeqVis/"},"https://vccri.github.io/SeqVis/")," "),l.a.createElement("li",null,l.a.createElement("a",{href:"tools/DVA_2.0.zip"},"DVA")," is a R script that performs differential variability analysis. A full description of this method can be found in this ",l.a.createElement("a",{href:"http://bioinformatics.oxfordjournals.org/content/24/13/i390.full"},l.a.createElement("i",null,"Bioinformatics")," paper"),"."),l.a.createElement("li",null,l.a.createElement("a",{href:"https://github.com/VCCRI/C3/"},"C3")," is an R package for cross-species compendium-based cell-type identification."),l.a.createElement("li",null,l.a.createElement("a",{href:"https://github.com/VCCRI/GEOracle"},"GEOracle")," is an R/shiny package for identifying and analysing perturbation experiments in NCBI GEO. Source code can be found in ",l.a.createElement("a",{href:"https://github.com/VCCRI/GEOracle"},"GitHub"),"."),l.a.createElement("li",null,l.a.createElement("a",{href:"https://github.com/VCCRI/Scavenger/"},"Scavenger")," a phyton tool to recover false negative unmapped reads."),l.a.createElement("li",null,l.a.createElement("a",{href:"http://bioinformatics.victorchang.edu.au/ularcirc.html"},"Ularcirc")," is an R package for visualisation and analysis of circular RNA."),l.a.createElement("li",null,l.a.createElement("a",{href:"https://github.com/VCCRI/CardiacProfileR"},"CardiacProfileR")," is an R package for extraction and visualisation of heart rate profiles from wearable fitness trackers."),l.a.createElement("li",null,l.a.createElement("a",{href:"https://github.com/VCCRI/starmap"},"starmap")," is a web-based VR-enabled tool which combines a 3D scatter plot with star plots to visualise hundreds of thousands of multivariate data points, such as single-cell expression data. starmap can be accessed from a desktop, laptop or a mobile device from the following link: ",l.a.createElement("a",{href:"https://vccri.github.io/starmap/"},"https://vccri.github.io/starmap/"),"."),l.a.createElement("li",null,l.a.createElement("a",{href:"https://github.com/VCCRI/FlowGrid"},"FlowGrid")," is an ultrafast clustering algorithm for single-cell flow cytometry data. It implements a novel grid-based DBSCAN clustering algorithm.")),l.a.createElement("h1",{className:"page-header"},"Data sets"),l.a.createElement("ul",null,l.a.createElement("li",null,"Antibody microarray of patients with Systemic Lupus Erythematosus (",l.a.createElement("a",{href:"http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?&acc=GSE27293"},"GEO"),")"),l.a.createElement("li",null,"Histone modification and gene expression in dosage compensation of ",l.a.createElement("i",null,"Drosophila melanogaster")," (",l.a.createElement("a",{href:"http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE34859"},"GEO"),")"),l.a.createElement("li",null,"Developmental timecourse of mouse ocular lens and whole embryo control (",l.a.createElement("a",{href:"http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE32334"},"GEO"),")"),l.a.createElement("li",null,"Gene expression during embryonic mouse molar tooth development (",l.a.createElement("a",{href:"http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE32321"},"GEO"),")"))))};var v=function(){return l.a.createElement("div",{id:"site_content"},l.a.createElement("div",{id:"content"},l.a.createElement("h1",null,"Contact"),l.a.createElement("p",null,"Dr. Joshua W. K. Ho",l.a.createElement("br",null),"Associate Professor ",l.a.createElement("br",null),"School of Biomedical Sciences, Li Ka Shing Faculty of Medicine",l.a.createElement("br",null),"Rm 4-44, Laboratory Block, 21 Sassoon Road",l.a.createElement("br",null),"Hong Kong SAR, China",l.a.createElement("br",null)),l.a.createElement("p",null,"Email: jwkho [ AT ] hku [ DOT ] hk ",l.a.createElement("br",null))))};var y=function(){return l.a.createElement("div",{id:"site_content"},l.a.createElement("div",{id:"content"},l.a.createElement("img",{src:"/img/JoshuaHo.jpg",width:"150",align:"right",alt:"",style:{margin:"10px 0px 0px 0px"}}),l.a.createElement("h1",null,"Dr. Joshua W. K. Ho"),l.a.createElement("h3",null,l.a.createElement("a",{href:"http://www.anatomy.hku.hk/staff/joshua-wing-kei-ho"},"Associate Professor, School of Biomedical Sciences")),l.a.createElement("div",null,l.a.createElement("h2",null," Education "),l.a.createElement("p",null,"2004-2006 ",l.a.createElement("b",null,"BSc (Hon 1, University Medal)"),", Biochemistry and Computer Science, The University of Sydney, Australia",l.a.createElement("br",null),"2007-2010 ",l.a.createElement("b",null,"PhD"),", Bioinformatics, The University of Sydney, Australia"),l.a.createElement("h2",null," Academic appointment "),l.a.createElement("p",null,"2010-2012 ",l.a.createElement("b",null,"Research Fellow in Medicine"),", Harvard Medical School and Brigham and Women's Hospital, Boston, USA ",l.a.createElement("br",null),"2012-2013 ",l.a.createElement("b",null,"Instructor in Medicine")," (Genetics), Harvard Medical School and Brigham and Women's Hospital, Boston, USA",l.a.createElement("br",null),"2013-2018 ",l.a.createElement("b",null,"Laboratory Head"),", Victor Chang Cardiac Research Institute, Australia",l.a.createElement("br",null),"2017-2018 ",l.a.createElement("b",null,"Division Head"),", Victor Chang Cardiac Research Institute, Australia",l.a.createElement("br",null))),l.a.createElement("div",null,l.a.createElement("h2",null," Research Interest"),l.a.createElement("ul",null,l.a.createElement("li",null,"Bioinformatics"),l.a.createElement("li",null,"Single-cell RNA-seq analysis"),l.a.createElement("li",null,"Gene regulation"),l.a.createElement("li",null,"Epigenomics"),l.a.createElement("li",null,"Systems biology"),l.a.createElement("li",null,"Artificial intelligence in medicine"),l.a.createElement("li",null,"Big data for biomedical sciences"))),l.a.createElement("div",null,l.a.createElement("h2",null," Thesis "),l.a.createElement("ol",null,l.a.createElement("li",null,l.a.createElement("b",null,"Ho JWK")," (2006) Computational Analysis of Evolution and Inference of Gene Regulatory Networks. BSc (Advanced) (Honours) Thesis, School of Biological Sciences, The University of Sydney"),l.a.createElement("li",null,l.a.createElement("b",null,"Ho JWK")," (2010) Principles and Applications of Bioinformatics for Systems Medicine. PhD Thesis, School of Information Technologies, The University of Sydney"))),l.a.createElement("div",null,l.a.createElement("h2",null," Award "),l.a.createElement("p",null,"2004-2005 Participated in the ",l.a.createElement("b",null,"Talented Student Program"),", The University of Sydney",l.a.createElement("br",null),"2006 ",l.a.createElement("b",null,"University Medal"),", The University of Sydney",l.a.createElement("br",null),"2007-2010 ",l.a.createElement("b",null,"Australia Postgraduate Award")," and ",l.a.createElement("b",null,"NICTA Research Project Award"),", The University of Sydney",l.a.createElement("br",null),"2009 ",l.a.createElement("b",null,"Best paper award"),", International Conference on Quality Software 2009",l.a.createElement("br",null),"2010-2013 ",l.a.createElement("b",null,"Ruth L. Kirschstein National Research Service Awards")," for an interdisciplinary postdoctoral fellowship at Harvard, National Institutes of Health, USA",l.a.createElement("br",null),"2014 ",l.a.createElement("b",null,"Rod Rickards Fellowship"),", Australian Academy of Science",l.a.createElement("br",null),"2015 ",l.a.createElement("b",null,"Illumina Early Career Researcher Award"),", Australian Epigenetics Alliance",l.a.createElement("br",null),"2015 ",l.a.createElement("b",null,"NSW Ministerial Award for Rising Stars in Cardiovascular Research"),", National Heart Foundation of Australia",l.a.createElement("br",null),"2016 ",l.a.createElement("b",null,"Young Tall Poppy Science Award"),", Australian Institute of Policy and Science",l.a.createElement("br",null),"2016 ",l.a.createElement("b",null,"Career Development Fellowship"),", National Health and Medical Research Council",l.a.createElement("br",null),"2016 ",l.a.createElement("b",null,"Future Leader Fellowship"),", National Heart Foundation of Australia",l.a.createElement("br",null),"2017 ",l.a.createElement("b",null,"Winner"),",",l.a.createElement("a",{href:"https://www.thinkable.org/vote_competitions/the-bioinformatics-peer-prize-ii"},"The Bioinformatics Peer Prize II"),", The Thinkable Academy"))))};var w=function(){var e=Object(n.useState)([]),a=Object(h.a)(e,2),t=a[0],r=a[1];return Object(n.useEffect)((function(){E.a.get("/data/recruitment.csv").then((function(e){var a=e.data;console.log(a);var t=b.a.parse(a,{header:!0});r(t.data)}))}),[]),l.a.createElement("div",{id:"site_content"},l.a.createElement("div",{id:"content"},l.a.createElement("h1",null,"Recruitment"),l.a.createElement("ul",null,t.map((function(e){return l.a.createElement("li",{key:e.ref_no},l.a.createElement("a",{href:e.url,target:"_blank",rel:"noopener noreferrer"}," ",e.title," ")," (Closing Date: ",e.closing_date,")")})))))};var S=function(){return l.a.createElement("div",{className:"App"},l.a.createElement("div",{id:"main"},l.a.createElement(c.a,null,l.a.createElement("div",null,l.a.createElement("div",{id:"header"},l.a.createElement("div",{id:"logo"},l.a.createElement("div",{id:"logo_text"},l.a.createElement("h1",null,l.a.createElement(c.b,{to:"/"},l.a.createElement("span",{className:"logo_colour"},"Bioinformatics and Systems Medicine"))),l.a.createElement("h2",null,"School of Biomedical Sciences, Li Ka Shing Faculty of Medicine, The University of Hong Kong"))),l.a.createElement("div",{id:"menubar"},l.a.createElement("ul",{id:"menu"},l.a.createElement("li",null,l.a.createElement(c.c,{to:"/",exact:!0,activeClassName:"selected"},"Home")),l.a.createElement("li",null,l.a.createElement(c.c,{to:"/people",exact:!0,activeClassName:"selected"},"People")),l.a.createElement("li",null,l.a.createElement(c.c,{to:"/research",exact:!0,activeClassName:"selected"},"Research")),l.a.createElement("li",null,l.a.createElement(c.c,{to:"/recruitment",exact:!0,activeClassName:"selected"},"Recruitment")),l.a.createElement("li",null,l.a.createElement(c.c,{to:"/publication",exact:!0,activeClassName:"selected"},"Publication")),l.a.createElement("li",null,l.a.createElement(c.c,{to:"/software",exact:!0,activeClassName:"selected"},"Software")),l.a.createElement("li",null,l.a.createElement(c.c,{to:"/contact",exact:!0,activeClassName:"selected"},"Contact"))))),l.a.createElement("div",{id:"content_header"}),l.a.createElement(o.c,null,l.a.createElement(o.a,{exact:!0,path:"/"},l.a.createElement(s,null)),l.a.createElement(o.a,{exact:!0,path:"/people"},l.a.createElement(m,null)),l.a.createElement(o.a,{exact:!0,path:"/people/JoshuaHo"},l.a.createElement(y,null)),l.a.createElement(o.a,{exact:!0,path:"/research"},l.a.createElement(u,null)),l.a.createElement(o.a,{exact:!0,path:"/recruitment"},l.a.createElement(w,null)),l.a.createElement(o.a,{exact:!0,path:"/publication"},l.a.createElement(f,null)),l.a.createElement(o.a,{exact:!0,path:"/software"},l.a.createElement(g,null)),l.a.createElement(o.a,{exact:!0,path:"/contact"},l.a.createElement(v,null))),l.a.createElement("div",{id:"content_footer"}),l.a.createElement("div",{id:"footer"},"Copyright \xa9 ",l.a.createElement("a",{href:"http://holab-hku.github.io/",target:"_blank",rel:"noopener noreferrer"},"Joshua Ho")),l.a.createElement("p",null,"\xa0")))))};Boolean("localhost"===window.location.hostname||"[::1]"===window.location.hostname||window.location.hostname.match(/^127(?:\.(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)){3}$/));i.a.render(l.a.createElement(l.a.StrictMode,null,l.a.createElement(S,null)),document.getElementById("root")),"serviceWorker"in navigator&&navigator.serviceWorker.ready.then((function(e){e.unregister()})).catch((function(e){console.error(e.message)}))}},[[29,1,2]]]);
//# sourceMappingURL=main.904acb74.chunk.js.map