--- 
title: "Introduction to R"
author: "Ho Lab"
date: "`r Sys.Date()`"

link-citations: yes
description: "This is an introduction of R."
output:
  html_document
---


# Prerequisites
## Installation
### Windows
**To Install R** 

- Open an internet browser and go to https://cran.r-project.org/.
- Click the "download R" link in the middle of the page under "Getting Started."
- Select a CRAN location (a mirror site) and click the corresponding link.
- Click on the "Download R for Windows" link at the top of the page.
- Choose the "base" and then Click on the "Download R 4.1.0 for Windows" link at the top of the page.
- Once the download is finished, you will obtain a file named "R-4.1.0-win.exe" or similar depending on the version of R that you download.
- Most of the time, you will likely want to go with the defaults, so click the button 'Next' until the process is complete.
- Now that R is installed, you need to download and install RStudio.

**To Install RStudio**

- Go to www.rstudio.com and click on the "Download RStudio" button.
- Click on "Download RStudio Desktop."
- The installation process is very straightforward as the gif below.

```{r fig1, echo=FALSE, message=FALSE, warning=FALSE}
knitr::include_graphics("./images/windows_rs.gif")
```

### Mac OSX
**To Install R**

- Open an internet browser and go to www.r-project.org.
- Click the "download R" link in the middle of the page under "Getting Started."
- Select a CRAN location (a mirror site) and click the corresponding link.
- Click on the "Download R for (Mac) OS X" link at the top of the page.
- Click on the file containing the latest version of R under "Files."
- Save the ``.pkg`` file, double-click it to open, and follow the installation instructions.
- Now that R is installed, you need to download and install RStudio.

**To Install RStudio**
Very similar with Windows,

- Go to www.rstudio.com and click on the "Download RStudio" button.
- Click on "Download RStudio Desktop."
- Click on the version recommended for your system, or the latest Mac version, save the ``.dmg`` file on your computer, double-click it to open, and then drag and drop it to your applications folder.
```{r fig2, echo=FALSE, message=FALSE, warning=FALSE}
knitr::include_graphics("./images/mac_rs.png")
```

### Ubuntu

**To Install R**

As it is common, prior to installing R, let us update the system package index and upgrade all our installed packages using the following two commands:
```
sudo apt update
sudo apt -y upgrade
```
After that, all that you have to do is run the following in the command line to install base R.
```
sudo apt -y install r-base
```

**To Install RStudio**

Once base R is installed, you can go ahead and install RStudio. For that we are going to head over again to the RStudio downloads page and download the ``.deb`` file for our Ubuntu version as shown in the image below:
```{r fig3, echo=FALSE, message=FALSE, warning=FALSE}
knitr::include_graphics("./images/ubuntu_rs.png")
```

## Use R inside RStudio
### R studio 
RStudio is a four pane work-space for 1) creating file containing R script, 2) typing R commands, 3) viewing command histories, 4) viewing plots and more.

Top-left panel: Code editor allowing you to create and open a file containing R script. The R script is where you keep a record of your work. R script can be created as follow:

`` File –> New –> R Script``

Bottom-left panel: R console for typing R commands

Top-right panel:
Workspace tab: shows the list of R objects you created during your R session
History tab: shows the history of all previous commands

Bottom-right panel:
Files tab: show files in your working directory
Plots tab: show the history of plots you created. From this tab, you can export a plot to a PDF or an image files
Packages tab: show external R packages available on your system. If checked, the package is loaded in R.
```{r fig4, echo=FALSE, message=FALSE, warning=FALSE}
knitr::include_graphics("./images/rstudio.png")
```

### Set working directory
-  Create a folder named “R-workshop” in your preferred directory
-  Create a "Data" folder in the "R-workshop" 
-  Download the data in the link https://drive.google.com/drive/folders/1RD6PY8IatnJ9ECcpAiCNZlc3KPBInGwo?usp=sharingused for this workshop and extract the data in the "Data" folder
-  From RStudio, use the menu to change your working directory under 
``Session > Set Working Directory > Choose Directory``
-  Choose the directory to “R-workshop”

```{r fig5, echo=FALSE, message=FALSE, warning=FALSE}
knitr::include_graphics("./images/directory.png")
```

Or you can type in the console:
```
setwd("/yourdirectory/R-workshop")
```
For Windows, the command might look like :
```
setwd("c:/yourdirectory/R-workshop")
```

### Install packages
There are two packages used in this workshop, in the R console, type:
```
install.packages('ggplot2')
install.packages('gplots')
```


# Data structure in R

R is a free and powerful statistical software for analyzing and visualizing data. In this chapter, we provide a quick and easy introduction to R programming. With the knowledge gained in this workshop, you will be ready to undertake your first very own data analysis. Oracle estimated over 2 million R users worldwide in 2012, cementing R as a leading programming language in statistics and data science. Every year, the number of R users grows by about 40%, and an increasing number of organizations are using it in their day-to-day activities. So, enjoy your R journey!


## Vector

Vectors are a fundamental concept in R, as many functions operate on and return vectors, so it is best to master these as soon as possible. For the technically inclined, in R, a (numeric) vector is an object consisting of a one-dimensional array of scalars.
HINT: YOU can get the documentation of a function using ?. E.g `?rep`

TRY:
```{r, results='hide'}
a <- rep(1,10)
```
```{r, results='hide'}
b <- 1:10
```
```{r, results='hide'}
c(a,b)
```
```{r, results='hide'}
a+b
```

**Quiz**
```
1. What is the difference of c(a,b) and a+b
2. Create a vector with "R" "is" "fun"
HINT: use c()
```
A new vector can be created by splicing an exisiting vector with a numerical indexes.
To produce a vector slice between two indexes, we can use the colon operator ``:``. This can be convenient for situations involving large vectors.
How to splice your vector?
```{r, results='hide'}
marks <- c(50, 100, 90, 80, 70) 
student_names <- c("Amy","Bobby","Cindy","Eddy","Dylon")
```
```{r, results='hide'}
names(marks) <- student_names
```
```{r, results='hide'}
marks[c(2, 3)] 
```
```{r, results='hide'}
marks[c("Bobby","Cindy")] 
```
**Quiz**
```
1. Get the highest mark amongst Amy, Cindy, Dylon
HINT: use max()
```

## Matrix

Matrix is a two dimensional data structure in R programming.

Matrix is similar to vector but additionally contains the dimension attribute. All attributes of an object can be checked with the ``attributes()`` function (dimension can be checked directly with the ``dim()`` function).

### Make a matrix

```{r, results='hide'}
m <- matrix(1:9, nrow = 3, dimnames = list(c("X","Y","Z"), c("A","B","C")))
#Change names
colnames(m) <- c("C1","C2","C3")
rownames(m) <- c("R1","R2","R3")
```
Make a matrix by using functions ``cbind()`` and ``rbind()`` as in column bind and row bind.
```{r, results='hide'}
cbind(c(1,2,3),c(4,5,6))
rbind(c(1,2,3),c(4,5,6))
```
```{r, results='hide'}
cbind(m,c(10,11,12))
rbind(m,c(10,11,12))
```
We can access elements of a matrix using the square bracket [ indexing method. Elements can be accessed as var[row, column]. Here rows and columns are vectors.
```{r, results='hide'}
m <- matrix(1:9, nrow = 3, dimnames = list(c("X","Y","Z"), c("A","B","C")))
# select rows 1 & 2 and columns 2 &
m[c(1,2),c(2,3)] 
# leaving column field blank will select entire columns
m[c(3,2),] 
```
### Modify a matrix 
Assign a value
```
-> ->>          rightwards assignment
<- <<-          assignment (right to left)
=               assignment (right to left)
```
Assign all elements less than 5 to 0
```{r, results='hide'}
m[m<5] <- 0
```
Transpose a matrix
```{r, results='hide'}
t(m)
```
Remove last row
```{r, results='hide'}
m <- m[-3,]
# or
m <- m[-nrow(m),]
```


## DataFrame 

Data frame is a two dimensional data structure in R. It is a special case of a list which has each component of equal length.
```{r, results='hide'}
x <- data.frame("SID" = 1:3, "Age" = c(23,25,21), "Name" = c("Amy","Bobby","Cindy"), "Mark" = c(100,82,75))
```
Accessing with ``[[`` or ``$`` is similar. However, it differs for ``[`` in that, indexing with ``[``  will return us a data frame but the other two will reduce it into a vector.
TRY:
```{r, results='hide'}
x["Name"]
x[,"Name",drop=F]
x[["Name"]]
x$Name
```
Select students with mark greater than 80 and make a new list called x_highmark.

```{r, results='hide'}
x_highmark <- x[x$Mark>80,]
```

**Quiz**
```
1.Select Students with mark greater than 80 and get the average age of them.
HINT: use mean()
```

## Factor
Factors are variables in R which take on a limited number of different values, such variables are often referred to as categorical variables.
```{r}
category = c(1,2,2,3,1,2,3,3,1,2,3,3,1)
fdata = factor(category)
fdata
```
Or you can factor the vector with specific names.
```{r}
fdata1 = factor(category,labels=c("A","B","C"))
fdata1
```
You can not treat the values in factors as numerical data, for example, if you use ``mean(fdata)`` you will get the warning as below.
```{r}
mean(fdata)
```
If we want to calculate the mean of the original numeric values of the fdata variable, we would have to convert the values using the levels function:
```{r}
mean(as.numeric(levels(fdata)[fdata]))
```
## List
List is a data structure having components of mixed data types. Lists are by far the most flexible data structure in R. They can be seen as a collection of elements without any restriction on the class, length or structure of each element. The only thing you need to take care of, is that you don't give two elements the same name. 
Data frames are a special case of list, but they have a few restrictions:
- you can't use the same name for two different variables
- all elements of a data frame are vectors
- all elements of a data frame have an equal length.

We can check if it’s a list with ``typeof()`` function and find its length using ``length()``.
```{r, results='hide'}
x <- list("a" = 2.5, "b" = TRUE, "c" = 1:3)
typeof(x)
length(x)
```
Compactly Display the Structure of an Arbitrary R Object using ``str()``
```{r, results='hide'}
str(x)
```

# Data manipulation
## apply functions
### apply()
``apply()`` takes Data frame or matrix as an input and gives output in vector, list or array. Apply function in R is primarily used to avoid explicit uses of loop constructs. It is the most basic of all collections can be used over a matrix.

```
apply(X, MARGIN, FUN)
Here:
-x: an array or matrix
-MARGIN:  take a value or range between 1 and 2 to define where to apply the function:
-MARGIN=1: the manipulation is performed on rows
-MARGIN=2: the manipulation is performed on columns
-MARGIN=c(1,2) the manipulation is performed on rows and columns
-FUN: tells which function to apply. Built functions like mean, median, sum, min, max and even user-defined functions can be applied>
```
Example1:
```{r, results='hide'}
m1 <- matrix(C<-(1:10),nrow=5, ncol=6)

m1_colsum <- apply(m1, 2, sum)
```
Example2: a function to find standard error was created, then passed into an apply function to find the standard error of each column.
```{r, results='hide'}
m2 <- matrix(c(1:10, 11:20, 21:30), nrow = 10, ncol = 3)

st.err <- function(x){
  sd(x)/sqrt(length(x))
}

apply(m2, 2, st.err)
```
Example3: the apply function is used to transform the values in each cell
If you set the MARGIN to 1:2 it will have the function operate on each cell.
```{r, results='hide'}
apply(m2, 1:2, function(x) x+3)
```
### lapply()
``lapply()`` in R takes list, vector or data frame as input and gives output in list.
```
lapply(X, FUN)
Arguments:
-X: A vector or an object
-FUN: Function applied to each element of x
```
Example:
```{r, results='hide'}
A<-c(1:9)
B<-c(1:12)
C<-c(1:15)
mylist<-list(A,B,C)
lapply(mylist, sum)
```
### sapply()
``sapply()`` works just like lapply, but will simplify the output if possible. This means that instead of returning a list like lapply, it will return a vector instead.
Example:
```{r, results='hide'}
sapply(mylist, sum)
```
### tapply()
Apply a function to each cell of a ragged array, that is to each (non-empty) group of values given by a unique combination of the levels of certain factors.

Example1:
```{r, results='hide'}
tdata <- as.data.frame(cbind(c(1,1,1,1,1,2,2,2,2,2), m2))
tapply(tdata$V2, tdata$V1, mean)
```
Example2:
```{r, results='hide'}
summary <- tapply(tdata$V2, tdata$V1, function(x) c(mean(x), sd(x)))
```
##  Pattern matching
### Find strings
``grep(pattern, string)`` returns by default a list of indices. If the regular expression, pattern, matches a particular element in the vector string, it returns the element's index.

For returning the actual matching element values, set the option value to TRUE by value=TRUE. Pattern matching in R defaults to be case sensitive. Turn the setting off with ignore.case = TRUE.(``tolower()`` and ``toupper()`` functions can convert everything to lower or upper case.)
```{r, results='hide'}
strings <- c("Abcd", "cdab", "ca bd")
grep("ab", strings)
grep("ab", strings, value = FALSE)
grep("ab", strings, value = TRUE)
grep("ab", strings, value = TRUE,ignore.case = TRUE)

```
### Find and replace patterns

``gsub(pattern, replacement, string)`` returns the modified string after replacing every pattern occurrence with replacement in string.
sub(pattern, replacement, string) replaces the first pattern occurrence.

```{r, results='hide'}
studentID <- c("u1000", "u1001", "u1002")
gsub("u", "U", studentID)

```

### Regular expression
**Basic**

Quantifiers specify how many times that the preceding pattern should occur.

- `*` : matches at least 0 times.
- `+` : matches at least 1 times.
- `?` : matches at most 1 times.
- `{n}` : matches exactly n times.
- `{n,}` : matches at least n times.
- `{,m}` : matches at most m times.
- `{n,m}` : matches between n and m times.
```{r, results='hide'}
strings <- c("a", "ab", "acb", "accb", "acccb", "accccb")
grep("ac*b", strings, value = TRUE)
grep("ac+b", strings, value = TRUE)
grep("ac?b", strings, value = TRUE)
grep("ac{2}b", strings, value = TRUE)
grep("ac{2,}b", strings, value = TRUE)
grep("ac{2,3}b", strings, value = TRUE)
```
- `^` : Start of the string.
- `$` : End of the string.
- `\\<` : Beginning of a word
- `\\>` : End of a word
```{r, results='hide'}
strings <- c("abcd", "cdab", "cabd", "c abd", "abc","*ab")
grep("^ab", strings, value = TRUE)
grep("ab$", strings, value = TRUE)
grep("\\<a", strings, value = TRUE)
grep("c\\>", strings, value = TRUE)
```
- `.` : Any non-empty character

- `[...]` : a permitted character list. Use `-` inside the brackets to specify a range of characters.

- `[^...]` : an excluded character list. Match any characters except those inside the square brackets.

- `|` : an OR operator, matches patterns on either side of the |.

- `\`: Escape metacharacters in regular expression, i.e.
```
$ * + . ? [ ] ^ { } | ( ) \`
```
```{r, results='hide'}
strings <- c("^ab", "ab", "abc", "abd", " abc d", "abe", "ab 12")
grep("ab.", strings, value = TRUE)
grep("ab[c-e]", strings, value = TRUE)
grep("ab[^c]", strings, value = TRUE)
grep("\\^ab", strings, value = TRUE)
grep("abc|abd", strings, value = TRUE)
```

**Advance**

- `[[:digit:]]` or `\\d` or `[0-9]` : digits 0 1 2 3 4 5 6 7 8 9
- `\\D` or `[^0-9]` : non-digits
- `[[:lower:]]` or `[a-z]` : lower-case letters

- `[[:upper:]]` or `[A-Z]` : upper-case letters

- `[[:alpha:]]` or `[A-z]` : alphabetic characters

- `[[:alnum:]]` or `[A-z0-9]` : alphanumeric characters

- `\\w` : word characters include alphanumeric characters `0-9,a-z,A-Z`, `-` and underscores `_`

- `\\W` : non-word characters

- `[[:blank:]]` : space and tab

- `[[:space:]]` or `\\s` : space characters: tab, newline, vertical tab, form feed, carriage return, space

- `\\S` : not space characters

- `[[:punct:]]` : punctuation characters

`! " # $ % & ' ( ) * + , - . / : ; < = > ? @ [  ] ^ _ ` { | } ~`

- `[[:cntrl:]]` or `\\c` : control characters, like ``\n`` or ``\r`` etc.

Regular expressions are a concise and flexible tool for describing patterns in strings. If you are interested, there are more examples in the tutorial provided by Cran.
https://cran.rstudio.com/web/packages/stringr/vignettes/regular-expressions.html

## Example

This exercise demonstrates how to use apply and tapply to summarize information from a table.
More information of apply functions can be found in *3.1*.

Using ``apply()`` to get the mean and sd on every column
```{r}
mtcars.mean <- apply(mtcars[,1:7], 2, mean)
mtcars.sd <- apply(mtcars[,1:7], 2, sd)
```
Using ``tapply()`` to get the maxmimum mpg per vs type
```{r}
mtcars.max.vs <- tapply(mtcars$mpg, mtcars$vs, max)
```
**Filtering**

Using ``grep()`` to select names with Merc
```{r}
mtcars[grep("Merc", rownames(mtcars)),]
```
combining ``tapply()`` and ``apply()``.
```{r}
apply(mtcars[,1:7],2,function(x) tapply(x, mtcars$gear, sum))
```
Using filter to select disp larger than 200.
```{r}
mtcars[mtcars$disp > 200,]
```

## Exercise

Load dataset mpg
```
library(ggplot2)
data("mpg")
```
***Quiz***

Q1 How  many different car manufacturers are present?
``Tip: Use functions: length() and unique()``

Q2 How many SUV's (class) does each car manufacturer have?
``Tip: Use table()``

Q3 Create a new data.table with only automatic transmission type (trans) and 4 cylinders (cyl)
``Tip: Use grepl()``

Q4 Create a new column called trans_subtype, with the transmission subtype information (the value inside the brackets)
``Tip: Use gsub``


Q5 For toyota's "toyota tacoma 4wd" model, rename them to only "tacoma 4wd"
``Tip: Try use gsub``


Q6 Create a new column called ID, with the first 3 characters of the manufacturer name capitalised with year of manufacturer separated by a _
``Tip: Use toupper() and substr() and paste0()
E.g. audi 1999 --> AUD_1999``



Q7 How many different model names have number(s) in their name for each manufacturer?
``Tips: Use table() and grepl()``

Q8 Find the average city mileage (cty) for each manufacturer type
``Tip: use sapply``

Q9 What is the average city mileage (cty) for models with number(s) in the name for each manufacturer?


Q10 Create a new data frame called mpg_summary with the following columns: 
- (A) Manufacturer name
- (B) number of different models for that manufacturer 
- (C) average displ
- (D) manufacturing range (e.g. 1999,2000,2003 --> 1999-2003)
- (E) average cyl
- (F) most frequent transmission type (if there is a tie, print out all separated by |)
- (G) least frequent drv (if there is a tie, print out all separated by |)
- (H) Max cty 
- (I) Min Hwy
- (J) All fuel types (fl). Separate by | (E.g. f,r,e --> "f|r|e")
- (K) Class with the longest name
- (L) Transition sub-type with the largest number (if there is tie, print out all separated by |)

``Tips: Use this to set up your new dataframe:``
```
mpg_summary <- data.frame(A=character(), B=character(), C=character(), D=character(), E=character(), F=character(), G=character(), H=character(),
I=character(), J=character(), K=character(), L=character())
```
Q11 Write a function that takes in a manufacturer name and print out the highway mileage (hwy) range for each car model



# Basic Statistics
Statistics is the study of how best to collect, analyze, and draw conclusions from data. A strong foundation will serve you well, no matter what industry you work in. This section describes basic (and not so basic) statistics.

## Generate data 
Height(cm) of 50 males and 50 females are estimated from https://dqydj.com/height-percentile-calculator-for-men-and-women/).
``set.seed(1) ``Set the seed to the random number generation to be 1. This ensures the results are reproducible across everyone's execution.
``rnorm()`` rnorm.X generates multivariate normal random variates.
```{r}
set.seed(1) 
male_height = rnorm(n = 50, mean = 175.6, sd = 11)
female_height = rnorm(n = 50, mean = 161.5, sd = 10.5)
```
We can keep only two decimal places using ``round()``.
```{r}
male_height = round(male_height, 2)
female_height = round(female_height, 2)
```
Check the number of elements in the vector.
```{r}
length(male_height)
```
Visualization of male group's height with histogram plot
```{r}
hist(male_height, freq = FALSE, main = "Histogram Plot")
```
Visualization of male group's height with density plot
```{r}
plot(density(male_height), main = "Density Plot")
```
Combine histogram plot and density plot using ``lines()``
The ``lines( )`` function adds information to a graph. It can not produce a graph on its own. Usually it follows a plot(x, y) command that produces a graph.
```{r}
hist(male_height, freq = FALSE, main = "Histogram and Density")
lines(density(male_height), col = "red")
```
Boxplot of height in male group and female group
```{r}
boxplot(male_height, female_height, names = c("male", "female"), ylab = "height(cm)")
```

Get basic information about the vector (minimum, maximum and quantiles) using ``summary()`` .
```{r}
summary(male_height)
summary(female_height)
```
Specific functions for Mean, Median, Minimum, Maximum, Standard deviation, Variance, Sum are shown below:
```
mean(), median(), min(), max(), sd(), var(), sum()
```


The R software provides access to the uniform distribution, ``runif()`` function generates random deviates of the uniform distribution and is written as ``runif(n, min = 0, max = 1)``. We may easily generate n number of random samples within any interval, defined by the min and the max argument.
Example: approximate the density function for X∼U(0,1)
```{r}
rand.unif <- runif(10000, min = 0, max = 1)
hist(rand.unif, freq = FALSE, xlab = 'x', density = 20)
```
If we have another record about weight for the same group of people, we can group weight and height data together into a list(assuming weight has correlation with height).
```{r}
set.seed(1)
male_weight = 0.52*male_height + runif(n = 50, min = -15, max = 15)
male_weight = round(male_weight, 2)
female_weight = 0.48*female_height + runif(n = 50, min = -14, max = 14)
female_weight = round(female_weight, 2)
```
Make a group Weight data and height data together.
```{r}
male_group = list(height = male_height, weight = male_weight)
female_group = list(height = female_height, weight = female_weight)
```
Use scatterplot to visualize the relationship of height and weight in male group
```{r}
plot(x = male_group$height,y = male_group$weight,
   xlab = "Height", ylab = "Weight",
   main = "Height vs Weight"
)
```


## Statistical analysis
Correlation: check the correlation between height and weight in male group and female group.
```{r}
cor(male_group$height, male_group$weight)
cor(female_group$height, female_group$weight)
```
``t.test()``: test if there is a significant difference between the means of height&weight in the male group and female group.
```{r}
t.test(male_group$height, female_group$height)
t.test(male_group$weight, female_group$weight)
```
``fisher.test()``: test the proportions of people with obesity in male group and  female group have significant difference or not.

Calculate BMI using equation BMI = weight(kg)/height(m)^2
```{r}
male_BMI = male_group$weight/(male_group$height/100)^2
female_BMI = female_group$weight/(female_group$height/100)^2
```
For people with BMI equal or larger than 30, define them as having obesity.
```{r}
male_obesity = ifelse(male_BMI >= 30, "Yes", "No")
female_obesity = ifelse(female_BMI >= 30, "Yes", "No")
```
Get the number of people with obesity in each group and summary in a matrix.
```{r}
male_num = table(male_obesity)
female_num = table(female_obesity)
obesity_num = rbind(male_num, female_num)
```
Fisher's exact test:
```{r}
fisher.test(obesity_num)
```


## Advanced plotting: ggplot2

[ggplot2](https://ggplot2.tidyverse.org/reference/index.html) is an R package to creat plots in a more flexible way. It can be used to generate different types of plots including `density plot`, `dot plot`, `box plot` and so on.

Here, we can use `ggplot2` to generate the scatterplot of height and weight in male group we did before.

First, we need to make a dataframe.

```{r}
library(ggplot2)
data <- data.frame(male_height, male_weight, female_height, female_weight)
```

```{r}
ggplot(data, aes(male_height, male_weight)) +
  geom_point() +
  labs(x = "Height", y = "Weight", title = "Height vs Weight")
```

[A violin plot](https://chartio.com/learn/charts/violin-plot-complete-guide/) depicts distributions of numeric data for one or more groups using density curves. The width of each curve corresponds with the approximate frequency of data points in each region.

Create data frame for violin plot including the height of the male group and the female group as well as a column indicating which group it comes from.

```{r}
data <- data.frame(height = c(male_height, female_height), group = c(rep("male", length(male_height)), rep("female", length(female_height))))
```

Use `ggplot2` to plot a violin plot of height in male and female group.

```{r}
ggplot(data, aes(factor(group),height)) +
  geom_violin() + geom_boxplot(width=0.1)+ 
  labs(x = "Group", y = "Height")
```


## Exercise

Use the PlantGrowth, perform t test to evaluate the weight of control and treatment groups.

**Quiz**

Q1 Extract weight of control and treatment groups and name them as ctrl_grp and trt1_grp.

Q2 Plot density for Control and Treatment1, with xlim from 2.5 to 7.5, and ylim from 0 tp 0.65.
HINT: use ``par(new = TRUE)`` to combine two graphs

Q3 Perform t test.


# RNA-seq analysis in R
This study examines the expression profiles of basal stem-cell enriched cells (B) and committed luminal cells (L) in the mammary gland of virgin, pregnant and lactating mice. Six groups are present, with one for each combination of cell type and mouse status. Each group contains two biological replicates.
## Read the sample information
The sampleinfo file contains basic information about the samples that we will need for the analysis today.
```{r}
# Read the sample information into R
sampleinfo <- read.table("data/SampleInfo.txt",header = TRUE)
View(sampleinfo)
sampleinfo
```

## Read the count data
You can use the ``head()`` command to see the first 6 lines. In RStudio the View command will open the dataframe in a new tab. The dim command will tell you how many rows and columns the data frame has.

```{r}
# Read the data into R
seqdata <- read.table("data/GSE60450_Lactation-GenewiseCounts.txt",header = TRUE, stringsAsFactors = FALSE)
View(seqdata)
dim(seqdata)
```

 The seqdata object contains information about genes (one gene per row), the first column has the Entrez gene id, the second has the gene length and the remaining columns contain information about the number of reads aligning to the gene in each experimental sample. There are two replicates for each cell type and timepoint.
Next, we should format the data and create a dataframe with 12 samples and 27179 genes.

## Format the data 

The first two columns in the seqdata dataframe contain annotation information. We need to make a new matrix containing only the counts, but we can store the gene identifiers (the EntrezGeneID column) as rownames. We will add more annotation information about each gene later on in the workshop.

Let’s create a new data object, countdata, that contains only the counts for the 12 samples.

```{r}
# Remove first two columns from seqdata
countdata <- seqdata[,-(1:2)]

# Store EntrezGeneID as rownames
rownames(countdata) <- seqdata[,1]
View(countdata)
```

As we can see the column names for the countdata now is V1 to V14, we want to have a more meaningful column names. But the sample name is too long so we want to just keep the first 7 characters.
```{r}
samplename <- substr(colnames(countdata), 1, 7)
samplename
```
And we can now modify the column names.
```{r}
colnames(countdata) <- samplename
head(countdata)
```

Now the column names are now the same as SampleName in the sampleinfo file. 

```{r}
table(colnames(countdata)==sampleinfo$SampleName)
```

## Reads per million (RPM)
Reads per million attempt to normalize for sequencing depth. Here’s how you do it for RPM:

- Count up the total reads in a sample and divide that number by 1,000,000 – this is our “per million” scaling factor.
- Divide the read counts by the “per million” scaling factor. This normalizes for sequencing depth, giving you reads per million (RPM)

```{r}
scaling_factor <- sum(countdata)/1000000
myRPM <- countdata/scaling_factor
```

## Filtering 

Summary of how many RPM greater than 0.5 there are in each row
```{r}
# There are 11433 genes that have RPM greater than 0.5 in all 12 samples.
table(rowSums(myRPM > 0.5))

# we would like to keep genes that have at least 2 RPM greater than 0.5 in each row of thresh
# Subset the rows of countdata to keep the more highly expressed genes
counts.keep <- countdata[rowSums(myRPM > 0.5) >= 2,]
dim(counts.keep)
```
**Normalization**
Count data is not normally distributed, so if we want to examine the distributions of the raw counts we need to log the counts. Next we’ll use box plots to check the distribution of the read counts on the natural logarithm scale. To avoid -infinite number of log(0) we manually assign log(0) to 0).
```{r}
logcounts <- as.matrix(log(counts.keep))
logcounts[is.infinite(logcounts)] <- 0
boxplot(logcounts, xlab="", ylab="Log counts per million",las=2)
```

## Select highly variable genes
Let’s select data for the 500 most variable genes and plot the heatmap.
```{r}
# We estimate the variance for each row in the logcounts matrix
var_genes <- apply(logcounts, 1, var)
head(var_genes)
```

Get the gene names for the top 500 most variable genes
```{r}
select_var <- names(sort(var_genes, decreasing=TRUE))[1:500]
head(select_var)
```

Subset logcounts matrix.
```{r}
highly_variable_lcpm <- logcounts[select_var,]
dim(highly_variable_lcpm)
```
## Hierarchical clustering with heatmaps

And finally we can do hierarchical clustering with heatmaps.``heatmap.2()`` is a very powerful tool from the ``gplots`` library to create enhanced heatmap with a dendrogram added to the left side and/or to the top.
Typically, reordering of the rows and columns according to some set of values (row or column means) within the restrictions imposed by the dendrogram is carried out.

This heatmap provides a number of extensions to the standard R ``heatmap`` function.
```{r, results='hide'}
library('gplots')
```
```{r}
heatmap.2(highly_variable_lcpm,trace="none", main="Top variable genes ",scale="row",cexCol=.8,cexRow=.8)

```

Create a folder called "result", and now we can save the heatmap into the "result" folder.

```{r, results='hide'}
png(file="./result/High_var_genes.heatmap.png",res=100)
heatmap.2(highly_variable_lcpm,trace="none", main="Top variable genes ",scale="row",cexCol=.8,cexRow=.8)
dev.off()
```
You can also use ``pdf()`` to save your figures in pdf format.

# Answers

## Exercise for Chapter2

```{r echo=TRUE}
library(ggplot2)
data("mpg")
```

A1 
```{r echo=TRUE}
length(unique(mpg$manufacturer))
```
A2
```{r echo=TRUE}
table(mpg[mpg$class=="suv","manufacturer"])
```
A3
```{r}
mpg[grepl("auto",mpg$trans)&mpg$cyl==4,]
```
A4
```{r}
mpg$trans_subtype <- gsub(".*\\((.*)\\)","\\1",mpg$trans)
```
A5
```{r}
mpg[mpg$model=="toyota tacoma 4wd","model"] <- "tacoma 4wd"
#or
mpg$model <- gsub("^toyota ", "",mpg$model)
```
A6
```{r}
mpg$ID <- paste0(toupper(substr(mpg$manufacturer,1,3)),"_",mpg$year)
```

A7
```{r}
table(mpg[grepl('\\d',mpg$model),"manufacturer"])
```
A8
```{r}
manufacturers <- unique(mpg$manufacturer)
sapply(manufacturers, function(x){
  mean(mpg[mpg$manufacturer==x,]$cty)
})
```
A9
```{r}
manufacturers <- unique(mpg$manufacturer)
sapply(manufacturers, function(x){
  x <- mean(mpg[mpg$manufacturer==x&grepl('\\d',mpg$model),]$cty)
  if(is.na(x)){0}else{x}
})
```
A10
```{r}
mpg_summary <- data.frame(A=character(), B=character(), C=character(), D=character(),
                          E=character(), F=character(), G=character(), H=character(),
                          I=character(), J=character(), K=character(), L=character())
for(x in unique(mpg$manufacturer)){
  subset <- mpg[mpg$manufacturer==x,]
  A <- x
  B <- length(unique(subset$model))
  C <- mean(subset$displ)
  D <- paste0(min(subset$year),"-",max(subset$year))
  E <- mean(subset$cyl)
  F <- paste(names(which(table(subset$trans)==max(table(subset$trans)))),collapse = "|")
  G <- paste(names(which(table(subset$trans)==min(table(subset$trans)))),collapse = "|")
  H <- max(subset$cty)
  I <- min(subset$hwy)
  J <- paste(unique(subset$fl),collapse = "|")
  K <- subset$class[which.max(nchar(subset$class))]

  largest_value <- max(gsub(".(\\d)","\\1",subset$trans_subtype[grepl("\\d",subset$trans_subtype)]))
  L <- paste(unique(subset$trans_subtype[grepl(largest_value,subset$trans_subtype)]), collapse = "|")
  mpg_summary[nrow(mpg_summary)+1,] <- c(A,B,C,D,E,F,G,H,I,J,K,L)
}
```
A11
```{r}
get_hwy_ranage <-function(x){
  lower <- min(mpg[mpg$manufacturer==x,"hwy"])
  upper <- max(mpg[mpg$manufacturer==x,"hwy"])
  print(paste0(lower,"-",upper))
}
```

## Exercise for Chapter3
```{r}
PlantGrowth
```


A1
```{r}
ctrl_grp <- PlantGrowth[PlantGrowth[, 2] == "ctrl",1]
trt1_grp <- PlantGrowth[PlantGrowth[, 2] == "trt1",1]
```

A2
```{r}
plot(density(ctrl_grp), main = "Control Group vs Treatment 1", xlab = "weight", xlim = c(2.5,7.5), ylim = c(0,0.65))
par(new = TRUE)
plot(density(trt1_grp), main = "", axes = FALSE, xlab = "", ylab = "", lty = 2, xlim = c(2.5,7.5), ylim = c(0,0.65))

```

A3
```{r}
t.test(ctrl_grp, trt1_grp)
```
