<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 3 Reference Genome | Falco Tutorial</title>
  <meta name="description" content="Falco: Single cell RNA-seq parrallelised cloud computing tutorial" />
  <meta name="generator" content="bookdown 0.16 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 3 Reference Genome | Falco Tutorial" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Falco: Single cell RNA-seq parrallelised cloud computing tutorial" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 3 Reference Genome | Falco Tutorial" />
  
  <meta name="twitter:description" content="Falco: Single cell RNA-seq parrallelised cloud computing tutorial" />
  

<meta name="author" content="Ho Lab" />


<meta name="date" content="2019-12-11" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="getting-started.html"/>
<link rel="next" href="read-data.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />












<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">A Minimal Book Example</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> About the Workshop</a></li>
<li class="chapter" data-level="2" data-path="getting-started.html"><a href="getting-started.html"><i class="fa fa-check"></i><b>2</b> Getting Started</a><ul>
<li class="chapter" data-level="2.1" data-path="getting-started.html"><a href="getting-started.html#prerequisites"><i class="fa fa-check"></i><b>2.1</b> Prerequisites</a></li>
<li class="chapter" data-level="2.2" data-path="getting-started.html"><a href="getting-started.html#aws-setup"><i class="fa fa-check"></i><b>2.2</b> AWS setup</a><ul>
<li class="chapter" data-level="2.2.1" data-path="getting-started.html"><a href="getting-started.html#aws-account"><i class="fa fa-check"></i><b>2.2.1</b> AWS account</a></li>
<li class="chapter" data-level="2.2.2" data-path="getting-started.html"><a href="getting-started.html#obtain-an-aws-secret-key"><i class="fa fa-check"></i><b>2.2.2</b> Obtain an AWS secret key</a></li>
<li class="chapter" data-level="2.2.3" data-path="getting-started.html"><a href="getting-started.html#install-the-aws-linux-client-aws-cli"><i class="fa fa-check"></i><b>2.2.3</b> Install the AWS Linux Client (AWS CLI)</a></li>
<li class="chapter" data-level="2.2.4" data-path="getting-started.html"><a href="getting-started.html#configuring-aws-linux-client"><i class="fa fa-check"></i><b>2.2.4</b> Configuring AWS Linux Client</a></li>
<li class="chapter" data-level="2.2.5" data-path="getting-started.html"><a href="getting-started.html#obtain-an-aws-ec2-key"><i class="fa fa-check"></i><b>2.2.5</b> Obtain an AWS EC2 key</a></li>
<li class="chapter" data-level="2.2.6" data-path="getting-started.html"><a href="getting-started.html#obtain-access-to-or-create-an-aws-s3-bucket"><i class="fa fa-check"></i><b>2.2.6</b> Obtain access to or create an AWS S3 bucket</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="reference-genome.html"><a href="reference-genome.html"><i class="fa fa-check"></i><b>3</b> Reference Genome</a><ul>
<li class="chapter" data-level="3.1" data-path="reference-genome.html"><a href="reference-genome.html#get-reference-genome-and-the-star-index"><i class="fa fa-check"></i><b>3.1</b> Get reference Genome and the STAR index</a></li>
<li class="chapter" data-level="3.2" data-path="reference-genome.html"><a href="reference-genome.html#sync-the-reference-genome-index-from-s3-bucket"><i class="fa fa-check"></i><b>3.2</b> Sync the reference genome index from s3 bucket</a></li>
<li class="chapter" data-level="3.3" data-path="reference-genome.html"><a href="reference-genome.html#check-reference-genome-index-from-s3-bucket"><i class="fa fa-check"></i><b>3.3</b> Check reference genome index from s3 bucket</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="read-data.html"><a href="read-data.html"><i class="fa fa-check"></i><b>4</b> Read Data</a><ul>
<li class="chapter" data-level="4.1" data-path="read-data.html"><a href="read-data.html#obtain-the-read-data"><i class="fa fa-check"></i><b>4.1</b> Obtain the Read Data</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="create-ami-for-falco-cluster-nodes.html"><a href="create-ami-for-falco-cluster-nodes.html"><i class="fa fa-check"></i><b>5</b> Create AMI for Falco cluster nodes</a><ul>
<li class="chapter" data-level="5.1" data-path="create-ami-for-falco-cluster-nodes.html"><a href="create-ami-for-falco-cluster-nodes.html#have-a-look-at-the-custom-ami-for-falco-framework-with-tools-pre-installed"><i class="fa fa-check"></i><b>5.1</b> Have a look at the Custom AMI for Falco framework with tools pre-installed</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="run-the-tutorial.html"><a href="run-the-tutorial.html"><i class="fa fa-check"></i><b>6</b> Run the tutorial</a><ul>
<li class="chapter" data-level="6.1" data-path="run-the-tutorial.html"><a href="run-the-tutorial.html#give-values-for-s3-bucket-and-user-name"><i class="fa fa-check"></i><b>6.1</b> Give values for S3 bucket and User Name</a></li>
<li class="chapter" data-level="6.2" data-path="run-the-tutorial.html"><a href="run-the-tutorial.html#launch-the-aws-emr-cluster"><i class="fa fa-check"></i><b>6.2</b> Launch the AWS EMR cluster</a></li>
<li class="chapter" data-level="6.3" data-path="run-the-tutorial.html"><a href="run-the-tutorial.html#monitor-the-emr-cluster"><i class="fa fa-check"></i><b>6.3</b> Monitor the EMR Cluster</a></li>
<li class="chapter" data-level="6.4" data-path="run-the-tutorial.html"><a href="run-the-tutorial.html#upload-the-manifest-file"><i class="fa fa-check"></i><b>6.4</b> Upload the Manifest file</a></li>
<li class="chapter" data-level="6.5" data-path="run-the-tutorial.html"><a href="run-the-tutorial.html#launch-the-split-job"><i class="fa fa-check"></i><b>6.5</b> Launch the Split job</a></li>
<li class="chapter" data-level="6.6" data-path="run-the-tutorial.html"><a href="run-the-tutorial.html#launch-the-pre-processing-job"><i class="fa fa-check"></i><b>6.6</b> Launch the Pre-processing job</a></li>
<li class="chapter" data-level="6.7" data-path="run-the-tutorial.html"><a href="run-the-tutorial.html#launch-the-analysis-job"><i class="fa fa-check"></i><b>6.7</b> Launch the Analysis job</a></li>
<li class="chapter" data-level="6.8" data-path="run-the-tutorial.html"><a href="run-the-tutorial.html#monitor-steps"><i class="fa fa-check"></i><b>6.8</b> Monitor Steps</a></li>
<li class="chapter" data-level="6.9" data-path="run-the-tutorial.html"><a href="run-the-tutorial.html#monitor-s3-files"><i class="fa fa-check"></i><b>6.9</b> Monitor S3 files</a></li>
<li class="chapter" data-level="6.10" data-path="run-the-tutorial.html"><a href="run-the-tutorial.html#download-results-and-terminate-cluster"><i class="fa fa-check"></i><b>6.10</b> Download results and Terminate Cluster</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Falco Tutorial</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="reference-genome" class="section level1">
<h1><span class="header-section-number">Chapter 3</span> Reference Genome</h1>
<div id="get-reference-genome-and-the-star-index" class="section level2">
<h2><span class="header-section-number">3.1</span> Get reference Genome and the STAR index</h2>
<p>üî¥‚ÄúBelow steps are not necessary in this workshop because we already save the files in the S3 bucket.‚Äù
In this tutorial, a human genome will be used. The file is ~800M in size.
In a work directory of your choice, create a genome directory, and download the files:</p>
<pre><code>mkdir genome_ref
cd genome_ref
wget http://hgdownload.cse.ucsc.edu/goldenPath/hg38/database/refFlat.txt.gz
wget ftp://ftp.sanger.ac.uk/pub/gencode/Gencode_human/release_21/GRCh38.genome.fa.gz
wget ftp://ftp.sanger.ac.uk/pub/gencode/Gencode_human/release_21/gencode.v21.chr_patch_hapl_scaff.annotation.gtf.gz
# unzip the files
gunzip *</code></pre>
<p>Download and install STAR
Before proceeding with this step, ensure that your Linux system has the following dependencies installed: make, gcc, g++ glibc-static.</p>
<p>In the work directory:</p>
<pre><code>wget -O STAR-2.5.2a.tar.gz https://github.com/alexdobin/STAR/archive/2.5.2a.tar.gz
tar -xzf STAR*.tar.gz
star_path=$( find . -name &quot;STAR&quot;|grep -E &quot;/Linux_x86_64/&quot; )
ln -s ${star_path%STAR} STAR</code></pre>
<p>Create the STAR index files</p>
<pre><code># create directory for star index
mkdir hg38_star_sparse_ref
STAR/STAR --runMode genomeGenerate --genomeDir hg38_star_sparse_ref/ --genomeFastaFiles genome_ref/GRCh38.genome.fa --genomeSAsparseD 2 --runThreadN 8 --sjdbGTFfile genome_ref/gencode.v21.chr_patch_hapl_scaff.annotation.gtf</code></pre>
<p>Copy the reference genome files and STAR index file to the AWS S3 bucket
Since Falco does not require the genome .fa file, you will only need to copy the two other reference files to the AWS S3 bucket. At a bash prompt from the genome_ref directory, execute the following command - first change [YOUR BUCKET] to your own bucket name:</p>
<pre><code>aws s3 sync . s3://[YOUR BUCKET]/falco-tutorial/genomes/hg38/genome_ref --exclude &quot;*.fa&quot; --exclude &quot;*.fa.gz&quot;
# change back to work directory
cd ..</code></pre>
<p>Copy the STAR index files to the AWS S3 bucket:</p>
<pre><code>aws s3 sync hg38_star_sparse_ref s3://[YOUR BUCKET]/falco-tutorial/genomes/hg38/star_index</code></pre>
</div>
<div id="sync-the-reference-genome-index-from-s3-bucket" class="section level2">
<h2><span class="header-section-number">3.2</span> Sync the reference genome index from s3 bucket</h2>
<p>In this step, the reference genome files and the STAR index files from our s3 bucket is synced to your s3 bucket.
The sync command is like below:</p>
<pre><code>#aws s3 sync s3://SOURCE-BUCKET-NAME s3://DESTINATION-BUCKET-NAME --source-region SOURCE-REGION-NAME --region DESTINATION-REGION-NAME

aws s3 sync s3://falcotutorial/genomes s3://[YOUR BUCKET]/falco-tutorial/genomes --source-region us-west-1 --region us-west-1</code></pre>
<p>Replace [YOUR BUCKET] with the name of your bucket.</p>
</div>
<div id="check-reference-genome-index-from-s3-bucket" class="section level2">
<h2><span class="header-section-number">3.3</span> Check reference genome index from s3 bucket</h2>
<p>The <code>star_index/</code> directory should look something like:</p>
<pre><code>chrLength.txt
chrNameLength.txt
chrName.txt
chrStart.txt
exonGeTrInfo.tab
exonInfo.tab
geneInfo.tab
Genome
genomeParameters.txt
SA
SAindex
sjdbInfo.txt
sjdbList.fromGTF.out.tab
sjdbList.out.tab
transcriptInfo.tab</code></pre>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="getting-started.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="read-data.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["Falco-tutorial.pdf", "Falco-tutorial.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
